### PR TITLE
NiFi-1481 Enhancement[ nifi.sh env]

### PR DESCRIPTION
**Pull request by PuspenduBanerjee**
Initial working version of nifi-1481 , with plenty of room for enhancement.
It may help to close https://issues.apache.org/jira/browse/NIFI-1481

### CODE REVIEW

**Comment by PuspenduBanerjee** [comment]
@apiri Can you please review it.

File: nifi-bootstrap/pom.xml [hunk 1]
```diff
@@ -41,5 +41,12 @@
 			<groupId>org.apache.nifi</groupId>
 			<artifactId>nifi-expression-language</artifactId>
 		</dependency>
+		<dependency>
+          <groupId>com.sun</groupId>
+          <artifactId>tools</artifactId>
+          <version>1.7.0+</version>
+          <scope>system</scope>
+          <systemPath>${java.home}/../lib/tools.jar</systemPath>
```
Comment (trkurc) [inline nifi-bootstrap/pom.xml]: I don't know if others have had this experience, but whenever tools was pulled in as a dependency this way, it tender to be fragile.

Comment (PuspenduBanerjee) [inline nifi-bootstrap/pom.xml]: Actually, we can remove that  declaration in pom. tools.jar/classes.jar has only runtime dependency.

**Comment by PuspenduBanerjee** [comment]
@trkurc Please review once you get a chance.

**Comment by PuspenduBanerjee** [comment]
@trkurc Any chance to review out of your busy schedule yet?

**Comment by trkurc** [comment]
I looked it over and gave it a try, I don't have any substantive comments yet, this was queued behind getting 0.5.1 out. I think this is a good candidate for 0.6.0.

**Comment by PuspenduBanerjee** [comment]
@trkurc Thanks. I shall keep it as is & look forward to see it as part of 0.6.0.

**Comment by apiri** [comment]
reviewing

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh [hunk 2]
```diff
@@ -118,6 +118,12 @@ locateJava() {
             fi
         fi
     fi
+    [ "x${TOOLS_JAR}" =  "x" ] && [ -n "${JAVA_HOME}" ] && TOOLS_JAR=$(find "${JAVA_HOME}" -name "tools.jar")
+    [ "x${TOOLS_JAR}" =  "x" ] && TOOLS_JAR=$(find "${JAVA_HOME}" -name "classes.jar")
+    if ["x${TOOLS_JAR}" =  "x" ]; then
```
Comment (apiri) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh]: missed a space after your [ in your if statement.

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/RunNiFi.java [hunk 3]
```diff
@@ -546,6 +552,31 @@ public void status() throws IOException {
...
 
+    public void env(){
+        final Logger logger = cmdLogger;
+        final Status status = getStatus(logger);
+        if(status.getPid() == null){
+            logger.info("Apache NiFi is not running");
+            return;
+        }
+        try{
+        final Class<?> virtualMachineClass=Class.forName("com.sun.tools.attach.VirtualMachine");
```
Comment (apiri) [inline nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/RunNiFi.java]: Formatting:  try block should be indented

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh (hunk 2 above)
Comment (apiri) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh]: I think my preference would be to opt to only display this if we are using the env command.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh [hunk 4]
```diff
@@ -118,6 +118,12 @@ locateJava() {
             fi
         fi
     fi
+    [ "x${TOOLS_JAR}" =  "x" ] && [ -n "${JAVA_HOME}" ] && TOOLS_JAR=$(find "${JAVA_HOME}" -name "tools.jar")
```
Comment (apiri) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh]: This was okay on OS X where my Java home was absolute, but not okay on Linux where this was a symlink.  Quick win here would be to use the -H of find

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/env-nifi.bat [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+rem    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+rem    See the License for the specific language governing permissions and
+rem    limitations under the License.
+rem
+
+rem Use JAVA_HOME if it's set; otherwise, just use java
+
+if "%JAVA_HOME%" == "" goto noJavaHome
+if not exist "%JAVA_HOME%\bin\java.exe" goto noJavaHome
+set JAVA_EXE=%JAVA_HOME%\bin\java.exe
```
Comment (apiri) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/env-nifi.bat]: Have to be careful with JAVA_HOME throughout as it is highly likely folks using Windows will have this located in a default location with spaces.  This caused troubles at points throughout.

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/RunNiFi.java [hunk 6]
```diff
@@ -546,6 +552,31 @@ public void status() throws IOException {
...
+        final Properties sysProps=(Properties)getSystemPropertiesMethod.invoke(virtualMachine);
+        for(Entry<Object, Object> syspropEntry: sysProps.entrySet()){
+            logger.info(syspropEntry.getKey().toString() + " = " +syspropEntry.getValue().toString());
+        }
+        Method detachMethod=virtualMachineClass.getDeclaredMethod("detach");
+        detachMethod.invoke(virtualMachine);
+        }catch(ClassNotFoundException cnfe){
+            logger.error("Seems tools.jar(Linux / Windows JDK) or classes.jar(Mac OS) is not available in classpath", cnfe);
+        }catch(Throwable t){
+            throw new RuntimeException(t);
```
Comment (apiri) [inline nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/RunNiFi.java]: Not sure the stacktrace provides much information worthwhile to the user.  Would opt to just leave the above log error and remove this.

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/RunNiFi.java [hunk 7]
```diff
@@ -546,6 +552,31 @@ public void status() throws IOException {
...
+        final Class<?> virtualMachineClass=Class.forName("com.sun.tools.attach.VirtualMachine");
+        Method attachMethod=virtualMachineClass.getMethod("attach", String.class);
+        Object virtualMachine=attachMethod.invoke(null, status.getPid());
+        Method getSystemPropertiesMethod=virtualMachine.getClass().getMethod("getSystemProperties");
+        final Properties sysProps=(Properties)getSystemPropertiesMethod.invoke(virtualMachine);
+        for(Entry<Object, Object> syspropEntry: sysProps.entrySet()){
+            logger.info(syspropEntry.getKey().toString() + " = " +syspropEntry.getValue().toString());
+        }
+        Method detachMethod=virtualMachineClass.getDeclaredMethod("detach");
+        detachMethod.invoke(virtualMachine);
```
Comment (apiri) [inline nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/RunNiFi.java]: detaching should likely happen in a finally block

**Comment by apiri** [comment]
@PuspenduBanerjee going to need to investigate a couple of things further.  With JAVA_HOME set, operation worked as anticipated for both OS X and Linux.  Windows had issues with the path declaration.  Quoted those items, but had issues with establishing NIFI_ROOT.  Cygwin also seemed to have

**Comment by PuspenduBanerjee** [comment]
@apiri For now I have no way to check Windows or Cygwin as I do not have any MS windows installation, total linux guy and on vacation to my homeland, so can't even borrow a windows PC from a colleague. . Please help to find out someone with Windows.

**Comment by trkurc** [comment]
@apiri @PuspenduBanerjee I can take a whack at those windows fixes.

**Comment by apiri** [comment]
Sounds good, @trkurc.  Just personally ran out of steam digging into it last evening.

**Comment by PuspenduBanerjee** [comment]
@trkurc Appreciate the Spirit.

**Comment by jvwing** [comment]
I have a [suggested fix/hack for dealing with spaces in the Java path in env-nifi.bat](https://github.com/jvwing/nifi/commit/2094080675d0d976fd4cd883d9f97cd10994ba69)).  The fix converts the TOOLS_JAR path into a short 8.3 path without spaces, so the eventual call to NiFi has only one quoted parameter.  The expanded TOOLS_JAR path was not working due to spaces, and two quoted paths make the whole thing blow up for reasons I do not comprehend.

However, this was not enough to get the env feature running on Windows 10.  Running it outputs 

```
17:58:32.959 [main] DEBUG o.a.n.b.NotificationServiceManager - Found 0 service elements
17:58:32.963 [main] INFO  o.a.n.b.NotificationServiceManager - Successfully loaded the following 0 services: []
17:58:32.965 [main] INFO  org.apache.nifi.bootstrap.RunNiFi - Registered no Notification Services for Notification Type NIFI_STARTED
17:58:32.987 [main] INFO  org.apache.nifi.bootstrap.RunNiFi - Registered no Notification Services for Notification Type NIFI_STOPPED
17:58:32.989 [main] INFO  org.apache.nifi.bootstrap.RunNiFi - Registered no Notification Services for Notification Type NIFI_DIED
17:58:33.012 [main] DEBUG org.apache.nifi.bootstrap.Command - Status File: bin\nifi.pid
17:58:33.013 [main] DEBUG org.apache.nifi.bootstrap.Command - Status File: bin\nifi.pid
17:58:33.040 [main] DEBUG org.apache.nifi.bootstrap.Command - Properties: {port=40930}
17:58:33.041 [main] DEBUG org.apache.nifi.bootstrap.Command - Pinging 40930
17:58:33.078 [main] DEBUG org.apache.nifi.bootstrap.Command - Sent PING command
17:58:33.080 [main] DEBUG org.apache.nifi.bootstrap.Command - PING response: PING
17:58:33.081 [main] INFO  org.apache.nifi.bootstrap.Command - Apache NiFi is not running
```

Where the last line "Apache NiFi is not running" matches RunNiFi.java line 559, where the env() method quits after not finding a PID.  Yes, NiFi was running when I tested this.  In comparison, the `status-nifi.bat` outputs

```
17:49:49.660 [main] DEBUG o.a.n.b.NotificationServiceManager - Found 0 service elements
17:49:49.664 [main] INFO  o.a.n.b.NotificationServiceManager - Successfully loaded the following 0 services: []
17:49:49.666 [main] INFO  org.apache.nifi.bootstrap.RunNiFi - Registered no Notification Services for Notification Type NIFI_STARTED
17:49:49.690 [main] INFO  org.apache.nifi.bootstrap.RunNiFi - Registered no Notification Services for Notification Type NIFI_STOPPED
17:49:49.691 [main] INFO  org.apache.nifi.bootstrap.RunNiFi - Registered no Notification Services for Notification Type NIFI_DIED
17:49:49.719 [main] DEBUG org.apache.nifi.bootstrap.Command - Status File: bin\nifi.pid
17:49:49.723 [main] DEBUG org.apache.nifi.bootstrap.Command - Status File: bin\nifi.pid
17:49:49.749 [main] DEBUG org.apache.nifi.bootstrap.Command - Properties: {port=40930}
17:49:49.752 [main] DEBUG org.apache.nifi.bootstrap.Command - Pinging 40930
17:49:49.798 [main] DEBUG org.apache.nifi.bootstrap.Command - Sent PING command
17:49:49.801 [main] DEBUG org.apache.nifi.bootstrap.Command - PING response: PING
17:49:49.804 [main] INFO  org.apache.nifi.bootstrap.Command - Apache NiFi is currently running, listening to Bootstrap on port 40930, PID=unknkown
```

Which recognizes that NiFi is running based on `isRespondingToPing()` (line 532), and describes the PID as "unknkown".  Since it looks like you use the PID in the env method, I'm not immediately sure what to suggest next.  It might be possible to work out an "if Windows, do this to get the PID" flow.

**Comment by trkurc** [comment]
I discovered the same issue with the pid on windows 10. My fix was to use quotes a bit differently on the cmd.exe line (cmd.exe /C ""%JAVA_EXE%" %JAVA_PARAMS% %BOOTSTRAP_ACTION%" ) and (set JAVA_PARAMS=-cp "%TOOLS_JAR%";%CONF_DIR%;%LIB_DIR%\* -Xms12m -Xmx24m %JAVA_ARGS% org.apache.nifi.bootstrap.RunNiFi)

**Comment by trkurc** [comment]
The pid is not being recorded in the nifi.pid file after I run RunNiFi with start or run, haven't been able to diagnose why yet. @markap14 may have an idea.

**Comment by trkurc** [comment]
push my changes up to a fork, in case anyone was interested (https://github.com/trkurc/nifi/commit/5ebf228a9ba5c5ce9f12189988920f94923e6210 from https://github.com/trkurc/nifi/commits/NIFI-1481)

**Comment by PuspenduBanerjee** [comment]
@trkurc  isn't `cmd.exe /C ""%JAVA_EXE%" %JAVA_PARAMS% %BOOTSTRAP_ACTION%" ` equivalent to `cmd.exe /C %JAVA_EXE%" %JAVA_PARAMS% %BOOTSTRAP_ACTION%" `

**Comment by trkurc** [comment]
@PuspenduBanerjee sadly, no, no it is not. https://technet.microsoft.com/en-us/library/bb490880.aspx

**Comment by trkurc** [comment]
(See the section labeled "Processing quotation marks ". ). So bizarre.

**Comment by PuspenduBanerjee** [comment]
@trkurc  Oh Man! Thanks for your guidance. Any chance of using power-shell in future?

**Comment by trkurc** [comment]
Also, @jvwing I fixed that misspelling of unknown when I was hacking on RunNiFi

**Comment by trkurc** [comment]
After an out of band discussion with @markap14, seems that windows pids might be challenging to get, so maybe we should leave the Windows env batch script out for 0.6.0. Any objections?

**Comment by trkurc** [comment]
@apiri - cygwin is proving to be a challenge, and not necessarily due to the changes in this patch. 

On my setup, the ':' separator on this line seems to break things (I suspect because my java is expecting ';' separators for things on the classpath)

See:
https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/nifi.sh#L192

**Comment by apiri** [comment]
@trkurc I think it might be a fair concession to have both this and Windows punted. Maybe we just roll in the check piggybacking off of the already existing $cygwin and figure out how we can improve later?

**Comment by trkurc** [comment]
@apiri: how would you think it best to review the changes I made based on your reviews? Another PR? A patch on the ticket?

**Comment by apiri** [comment]
Patch to apply on top of the PR is probably simplest, but am good with whatever is easiest for you.

**Comment by trkurc** [comment]
@apiri - I attached a patch to the ticket, which should be the last two commits off my branch https://github.com/trkurc/nifi/commits/NIFI-1481

**Comment by apiri** [comment]
@trkurc Okay, will apply over the top and review.  Thanks!

**Comment by apiri** [comment]
@trkurc Patch unfortunately is not applying even on a fresh checkout of the 218 PR.  Will checkout your branch and work from there.

Patch issue was:

> $ git am --signoff ~/Downloads/NIFI-1481.001.patch
Applying: NIFI-1481: Code review fixes
error: patch failed: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/env-nifi.bat:50
error: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/bin/env-nifi.bat: patch does not apply
Patch failed at 0001 NIFI-1481: Code review fixes

**Comment by apiri** [comment]
@trkurc What is the final verdict on handling of Windows and Cygwin?  Are we punting those for now? If so, would like to provide handling for env with cygwin in nifi.sh and removal of the .bat for now. Happy to do these myself, but wanted to avoid another level of indirection.  Let me know what works.

Otherwise, your branch as presented looks good and works as anticipated.

**Comment by trkurc** [comment]
@apiri - I tried last night to get the pid to get Windows to work. I could not. I say we shelve the env batch script (remove it for now from the patch set). 

On cygwin does "nifi.sh run" or "nifi.sh status" work at all for you without these patches? It did not for me. What were you proposing to fix cygwin? Get run and start working, then try to get env working as well?

**Comment by trkurc** [comment]
So, after discussion with @apiri, we removed the env batch script because it did not work due to limitations of bootstrap, and reworked nifi.sh to build the classpath for Bootstrap to work with cygwin and !cygwin and only include TOOLS_JAR if set.

**Pull request closed without merging**